File: Swift/0004.swift
class Solution {
    func findMedianSortedArrays(_ nums1: [Int], _ nums2: [Int]) -> Double {
        let A = nums1
        let B = nums2
        let m = A.count
        let n = B.count
        
        // Ensure that A is the smaller array
        if m > n {
            return findMedianSortedArrays(B, A)
        }
        
        var imin = 0
        var imax = m
        let halfLen = (m + n + 1) / 2
        
        while imin <= imax {
            let i = (imin + imax) / 2
            let j = halfLen - i
            
            if i < m && B[j - 1] > A[i] {
                imin = i + 1
            } else if i > 0 && A[i - 1] > B[j] {
                imax = i - 1
            } else {
                let maxLeftA = (i == 0) ? Int.min : A[i - 1]
                let minRightA = (i == m) ? Int.max : A[i]
                let maxLeftB = (j == 0) ? Int.min : B[j - 1]
                let minRightB = (j == n) ? Int.max : B[j]
                
                if (m + n) % 2 == 1 {
                    return Double(max(maxLeftA, maxLeftB))
                } else {
                    return (Double(max(maxLeftA, maxLeftB)) + Double(min(minRightA, minRightB))) / 2.0
                }
            }
        }
        
        fatalError("Input arrays are not sorted or of incorrect length.")
    }
}

File: Dart/0004.dart
import 'dart:math';

class Solution {
  double findMedianSortedArrays(List<int> nums1, List<int> nums2) {
    // Ensure nums1 is the smaller array
    if (nums1.length > nums2.length) {
      return findMedianSortedArrays(nums2, nums1);
    }

    int m = nums1.length;
    int n = nums2.length;
    int imin = 0;
    int imax = m;
    double halfLen = (m + n + 1) / 2.0;

    while (imin <= imax) {
      int i = (imin + imax) ~/ 2;
      int j = (halfLen - i).toInt();

      // Partition values
      double maxX = i == 0 ? double.negativeInfinity : nums1[i - 1].toDouble();
      double minX = i == m ? double.infinity : nums1[i].toDouble();
      double maxY = j == 0 ? double.negativeInfinity : nums2[j - 1].toDouble();
      double minY = j == n ? double.infinity : nums2[j].toDouble();

      if (maxX <= minY && maxY <= minX) {
        if ((m + n) % 2 == 0) {
          return (max(maxX, maxY) + min(minX, minY)) / 2.0;
        } else {
          return max(maxX, maxY);
        }
      } else if (maxX > minY) {
        imax = i - 1;
      } else {
        imin = i + 1;
      }
    }

    throw ArgumentError("Input arrays are not sorted or invalid");
  }
}
